Gutenboarding: Add passwordless signup form to `/gutenboarding…

roo2 opened the pull request:
WIP, not ready for review yet, see issue https://github.com/Automattic/wp-calypso/issues/38792

#### Changes proposed in this Pull Request

* Building on top of work that @jameslnewell is doing to pull out a shared authentication package
* Will have to discuss how to do translation/ i18n
* Need to wait for the authentication package to be merged

<img width="451" alt="image" src="https://user-images.githubusercontent.com/1500769/72944469-d3e09900-3ddd-11ea-91d2-c984aec66c08.png">

#### Testing instructions

* Use an incognito window, otherwise you're going to get logged out of your other tabs
* `/gutenboarding/signup`
* Enter an email that will succeed
* New user info should be shown on screen, verify it looks correct
* Enter an email that will fail
* API response will be shown on screen, verify it looks correct

Fixes #38792

Commit:
WIP: first draft of gutenboarding signup form

simison commented on client/landing/gutenboarding/onboarding-block/signup-form/index.tsx:
```diff
@@ -0,0 +1,68 @@
+
+	const renderTosLink = () => {
+		return (
+			<a href="https://wordpress.com/tos/" target="_blank" rel="noopener noreferrer">
```
[`ExternalLink`](https://github.com/WordPress/gutenberg/tree/8e7e67bae2d712120137ca86977c10994e8c637a/packages/components/src/external-link#readme) component might be useful here.

simison commented on client/landing/gutenboarding/onboarding-block/signup-form/index.tsx:
```diff
@@ -0,0 +1,68 @@
+					onChange={ setEmailVal }
+					placeholder="yourname@email.com"
+				/>
+				<p>By creating an account you agree to our { renderTosLink() }.</p>
```
This one will be a fun one to try apply `__experimentalCreateInterpolateElement` from `@wordpress/i18n` :-)

roo2 replied:
It turns out that the` __experimentalCreateInterpolateElement` in `@wordpress/i18n` isn't in the version of the package that we're using (3.4.0). We had that card to update the packages to the latest version but I couldn't find it using the issue search but I think we will have to do that first.

simison replied:
>We had that card to update the packages to the latest version but I couldn't find it using the issue search but I think we will have to do that first.

This was the latest I saw go in https://github.com/Automattic/wp-calypso/pull/38554

simison replied:
It might make sense to just format this paragraph differently for now to get this merged and update the copy later.

p-jackson replied:
`__experimentalCreateInterpolateElement` is available now :) but it's not in the typings :(

It works well so I'll stick with it, but I'll look for some way to hack it in to the types of our project before submitting something to the official repo.

roo2 committed:
hacky styling to look like gutenberg form

roo2 committed:
make signup form look more like gutenberg

roo2 committed:
make gutenboarding signup bigger

p-jackson commented:
I switched the PR's base branch to make it clearer which are your changes, hope that's ok feel free to change it to what you think's appropriate @roo2

p-jackson commented on client/landing/gutenboarding/onboarding-block/signup-form/index.tsx:
```diff
@@ -0,0 +1,80 @@
+	Client,
+} from '@automattic/authentication';
+import { Button, TextControl, Modal } from '@wordpress/components';
+import FormLabel from 'components/forms/form-label';
```
Rather than pulling in `<FormLabel>` from Calypso, `<TextControl>` comes with it's own `label` prop. It looks like this rendered:

<img width="278" alt="image" src="https://user-images.githubusercontent.com/1500769/72690063-51f23500-3b7d-11ea-9aa9-c3c60e957dfb.png">

p-jackson commented on client/landing/gutenboarding/onboarding-block/signup-form/index.tsx:
```diff
@@ -0,0 +1,80 @@
+					value={ emailVal }
+					onChange={ setEmailVal }
+					placeholder="yourname@email.com"
+				/>
```
Maybe add a `type="email"` prop here so that the correct keyboard appears on tablets?

p-jackson commented on client/landing/gutenboarding/onboarding-block/signup-form/index.tsx:
```diff
@@ -0,0 +1,80 @@
+	};
+
+	return (
+		<Modal className="signup-form" title="Sign up to save your changes" onRequestClose={ () => {} }>
```
The lint on this line is bothersome (I don't see what the problem is with empty lambdas). But you can silence it by using the `noop` utility from `lodash`.

roo2 committed:
apply config and use translations

roo2 committed:
disable button on submit

ramonjd commented:
Nice work @roo2 

@jameslnewell What's the status of the `feat/authentication-package` branch? Is the anything we can do to coax it along? Bribes? Flattery? Kudos broadside?

ramonjd commented:
> What's the status of the feat/authentication-package branch? Is the anything we can do to coax it along? Bribes? Flattery? Kudos broadside?

Actually, based on our conversation in p1579487677022600-slack-CRWCHQGUB we could use this PR without the [@automattic/authentication handlers](https://github.com/Automattic/wp-calypso/pull/38882/files#diff-f294baa8838ad36a5984a11fcc2b96c0R27) and build on top of @andrewserong 's work in https://github.com/Automattic/wp-calypso/pull/38823 

We _might_ need some customization anyway, and it'll take the pressure off Team Calypso.
 
If it turns that we require @automattic/authentication for the API request stuff later we can use it. We'll still need the user state layer and so on.

ramonjd commented on client/landing/gutenboarding/onboarding-block/signup-form/index.tsx:
```diff
@@ -0,0 +1,86 @@
+
+	const handleSignUp = ( event: React.FormEvent< HTMLFormElement > ) => {
+		event.preventDefault();
+		signUp( emailVal );
```
Once the work in https://github.com/Automattic/wp-calypso/pull/38823/files#diff-5fdbca26e3ba6bb6ea1026beb3049c5fR76 is ready, we can swap out the method and this will be golden.

roo2 commented:
Thanks for the feedback guys! I am going to leave this task alone for a little bit until the user store and proxy work that Andy has done gets merged https://github.com/Automattic/wp-calypso/pull/38823, and hopefully the `automattic/authentication` work is reconciled/merged as well. And then I'll come back to it and finish this :)

p-jackson committed:
Signup using User data-store functions instead of authentication package

p-jackson committed:
Use isDismissible to remove close button from signup modal for now

@wordpress/components typings have a typo, the isDismissable (with and
'a') spelling doesn't work.

Use lodash.noop to work around empty lambda lint warning.

p-jackson committed:
Show button animation while loading

p-jackson committed:
Use TextControl label to label email field

This way we don't need to import any controls from Calypso

p-jackson committed:
Add translation context for email placeholder

p-jackson committed:
Add type=email to email field

Adds client-side validation and will show the correct keyboard layout
on mobile devices.

p-jackson committed:
Translate ToS link using __experimentalCreateInterpolateElement

p-jackson committed:
Show new user on screen for easier testing

ramonjd commented on client/landing/gutenboarding/onboarding-block/signup-form/index.tsx:
```diff
@@ -0,0 +1,78 @@
+	return (
+		<Modal
+			className="signup-form"
+			isDismissible={ false }
```
Should be `isDismissable`

p-jackson replied:
[Reading the GB source](https://github.com/WordPress/gutenberg/blob/c079120ce39044e53ea41a6a1a246aeec8195433/packages/components/src/modal/index.js#L113), `dismissable` is deprecated. It also doesn't hide the close button, which is what I was trying to achieve.

Unfortunately the TypeScript types don't line up with this so I'll push an update to those too p1579736961026000-slack-C45SNKV4Z.

ramonjd commented on client/landing/gutenboarding/onboarding-block/signup-form/index.tsx:
```diff
@@ -0,0 +1,78 @@
+					<Button
+						type="submit"
+						className="signup-form__submit"
+						disabled={ isFetchingNewUser }
```
Could we also disable this when `! emailVal` 

Something similar to what we do in https://github.com/Automattic/wp-calypso/blob/c9e127d/client/blocks/signup-form/passwordless.jsx

p-jackson replied:
I personally I like setting `type=email` on the input box and then letting default browser behaviour kick in, that way we know the error message will be accessible. Usually I think the browser won't tell you the email is invalid until you attempt to submit.

Maybe it can be in another PR after we have more of an idea of how passwordless signup should look and feel in gutenboarding. Will ping @Automattic/gutenboarding-design so they can weigh in/think about how invalid user input should work in this brave new world.

ramonjd commented on client/landing/gutenboarding/onboarding-block/edit.tsx:
```diff
@@ -57,6 +58,9 @@ const OnboardingEdit: FunctionComponent< BlockEditProps< Attributes > > = () =>
 					{ ! siteVertical ? <Redirect to={ Step.IntentGathering } /> : <DesignSelector /> }
 				</Route>
+				<Route exact path={ Step.Signup }>
+					<SignupForm />
```
Note for future, related PR: 

Create an action to show modal, so we can flip it open at any point in the journey

ramonjd commented:
This is working well. Thanks @roo2 and @p-jackson !!

I'll leave it open for other reviewers, but I think for an initial implementation it's good to go. 

We can optimize/update later when we've integrated the sites request and we're implementing the actual flow.

andrewserong commented on client/landing/gutenboarding/onboarding-block/signup-form/index.tsx:
```diff
@@ -0,0 +1,78 @@
+/**
+ * External dependencies
+ */
+import { noop } from 'lodash';
```
I don't think we've used any lodash in Gutenboarding so far. Since it sounds like the ultimate goal is to move Gutenboarding into `/apps` (where we might not end up using lodash), would it make sense to replace `noop` with `() => undefined`?

ramonjd replied:
I thought this too, but we do use a few (`times`, `map` and `partition`). 

Agree that `noop` isn't really required.

andrewserong replied:
Ah, I missed that in my search! Thanks.

p-jackson replied:
I don't wanna start pulling in `lodash` if we don't have to! I feel like modern JS has enough goodies in it without the utility library. I really only pulled it in so I could shut up an eslint warning. Maybe `() => undefined` will do the trick though ...

p-jackson replied:
It does! Thanks for the tip @andrewserong 👍

sirreal replied:
https://github.com/Automattic/wp-calypso/pull/37713#discussion_r348023578 😏

andrewserong commented:
This is looking good to me, too! Just left a question about lodash, but other than that, once the failing typecheck is fixed, I think we can merge it in and then iterate on it.

ramonjd commented on client/landing/gutenboarding/stores/user/index.ts:
```diff
@@ -0,0 +1,14 @@
+ */
+import config from '../../../../config';
+
+export const USER_STORE = User.register( {
```
We'll need this for future requests. Is it something we could add to `client/landing/gutenboarding/constants.ts` or somewhere more appropriate?

p-jackson replied:
Don't like the idea of initialising runtime guff (like registering stores) in `constants.ts`. Maybe we should just change the data-store so it exports the key? What do you think about that @andrewserong? Was there a reason you wanted `.register` to return the key? It would mean we could do things like assign a random store key to avoid collisions?

p-jackson replied:
To be clear, if we exported the key from the data-store then the consuming code in a component might look like this:

```typescript
import { USER_STORE_KEY } from '@automattic/data-stores';

const { createAccount } = useDispatch( USER_STORE_KEY );
```

andrewserong replied:
Good question! I was mostly following the pattern used by the other stores. What I like about the pattern is that the consuming code is responsible for registering the store, instead of the data store package. So you can import anything from `packages/data-stores` but nothing gets registered until you specifically request it. The other thing we discussed was keeping the client id and secret out of `packages/data-stores` and `User.register( /* keys */ );` seemed like a natural place to do it. I'm open to other ideas, though!

Since the user store is going to be used potentially in a bunch of places, would it make sense to think of it as a part of the init process of loading Gutenboarding? So it could happen somewhere around `Window.AppBoot` in `client/landing/gutenboarding/index.tsx`? There's possibly a nice parallel there with `client/landing/login/index.js` which runs `const user = userFactory();` in `window.AppBoot`.

andrewserong replied:
Oh, wait, I might have misunderstood, we're just talking about the constant for the store key, aren't we? In that case, I think it makes sense to import it straight from `@automattic/data-stores`, but I'm wondering if the pattern used in the other packages was to ensure that you can only get the constant _after_ the store has been registered?

p-jackson replied:
Yeah I definitely like that the package doesn't have the side-effect of registering a store, just because it gets imported. I was thinking about make an `initStores()` function that gets called in `AppBoot`. Might be nicer that way because then the way and order that everything gets init'd will be deterministic.
It'll also give us a way to freshly init the stores in unit tests. So maybe I'll do that.

It's a toss up whether to export the store key from `@automattic/data-stores`. On the one hand I like the data hiding of having `register` return the key; other code isn't supposed to just know what that constant is. On the other hand the looser coupling of looking up a store at runtime using its key seems to be a feature of `@wordpress/data`, like it was designed that way i.e. you're supposed to be able to look up a store by its constant and fail gracefully if no one has registered it yet.

I'm inclined to just leave `@automattic/data-stores` the way it is since I don't know which is the right trade off for us.

andrewserong replied:
> Yeah I definitely like that the package doesn't have the side-effect of registering a store, just because it gets imported. I was thinking about make an initStores() function that gets called in AppBoot.

I like this idea, sounds neat and could be a helpful way of tracking the number of stores we're using in one place, too.

> I'm inclined to just leave @automattic/data-stores the way it is since I don't know which is the right trade off for us.

I'm not sure of the right trade off here either... I think the answer might become clearer to us once we have more use cases in Gutenboarding to guide us :)

p-jackson committed:
Remove lodash

p-jackson commented:
> once the failing typecheck is fixed, I think we can merge it in and then iterate on it.

I was actually thinking of merging and fixing typescript errors later. We could just hack the typings into our project, but since it seems like it'd be best to fix them would be in the upstream packages too, I thought they seemed harmless enough to just keep around until they got deployed upstream.

... although it will put a red check beside every PR we open until they get deployed ... so maybe it would be worth hacking in a fix in our project initially ...

andrewserong commented:
> I was actually thinking of merging and fixing typescript errors later. We could just hack the typings into our project, but since it seems like it'd be best to fix them would be in the upstream packages too, I thought they seemed harmless enough to just keep around until they got deployed upstream.

I was thinking of this too (and it feels pretty harmless to me, too), but I was wondering if it might be disconcerting for other folks, because it would then put a red `x` on commits on master due to the `typecheck-strict` circle CI job? 🤔

andrewserong commented:
To get the CI passing, can we add a couple of `@ts-ignore` lines for the bits that we should circle back to and fix up later?

p-jackson commented:
> To get the CI passing, can we add a couple of @ts-ignore lines for the bits that we should circle back to and fix up later?

Yeah that's probably the easier hackery to get it going.

FYI: DefinitelyTyped/DefinitelyTyped#41800

p-jackson committed:
Silence TypeScript errors

p-jackson commented:
Our lint settings are pretty strict and `@ts-ignore` isn't allowed. The type casting shenanigans in 699f845 should do the trick until the issues are fixed upstream.

andrewserong reviewed (approved):
Nice work on this @p-jackson and @roo2! And thanks for submitting the upstream change to [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/41800), Phil!

I think it's fine for us to iterate on where we're importing / registering the User store in future PRs when we start to use it in other places.

sirreal commented on client/landing/gutenboarding/onboarding-block/signup-form/index.tsx:
```diff
@@ -72,4 +64,13 @@ const SignupForm = () => {
+			link_to_tos: <ExternalLink href="https://wordpress.com/tos/" />,
+		}
+	);
+}
```
I see we had a function before, but is there a reason to wrap this in a function call?

Can we do:

```tsx
const tosLink = __experimentalCreateInterpolateElement( … );
// …
return <Modal>{ tosLink }</Modal>;
```

right in the component body instead?

p-jackson replied:
> is there a reason to wrap this in a function call?

Just personal preference probably. I could argue that it makes the component's function body shorter; but it doesn't hide any branching so it's not removing any complexity from the function.

One of the (small) cons I see in hooks is that component function bodies can get long, so I'm always looking for ways to split things out if I can come up with a good name for it.

Hooks also bring a new way to write abstractions, so you can use them to shorten functions too, so it goes both ways :)
